aiida/engine/processes/workchains/restart.py

```diff
@@ -0,0 +1,423 @@
+        self.ctx.process_name = self._process_class.__name__
+        self.ctx.unhandled_failure = False
+        self.ctx.is_finished = False
+        self.ctx.iteration = 0
+
+    def should_run_process(self):
+        """Return whether a new process should be run.
+
+        This is the case as long as the last process has not finished successfully and the maximum number of restarts
+        has not yet been exceeded.
+        """
+        return not self.ctx.is_finished and self.ctx.iteration < self.inputs.max_iterations.value
+
+    def run_process(self):
+        """Run the next process, taking the input dictionary from the context at `self.ctx.inputs`."""
+        self.ctx.iteration += 1
+
+        try:
+            unwrapped_inputs = self.ctx.inputs
+        except AttributeError:
+            raise AttributeError('no process input dictionary was defined in `self.ctx.inputs`')
+
+        # Set the `CALL` link label
+        unwrapped_inputs['metadata']['call_link_label'] = 'iteration_{:02d}'.format(self.ctx.iteration)
+
+        inputs = self._wrap_bare_dict_inputs(self._process_class.spec().inputs, unwrapped_inputs)
+        node = self.submit(self._process_class, **inputs)
+
+        # Add a new empty list to the `errors_handled` extra. If any errors handled registered through the
+        # `register_error_handler` decorator return an `ErrorHandlerReport`, their name will be appended to that list.
+        errors_handled = self.node.get_extra('errors_handled', [])
+        errors_handled.append([])
+        self.node.set_extra('errors_handled', errors_handled)
+
+        self.report('launching {}<{}> iteration #{}'.format(self.ctx.process_name, node.pk, self.ctx.iteration))
+
+        return ToContext(children=append_(node))
+
+    def inspect_process(self):
+        """Analyse the results of the previous process and call the error handlers when necessary."""
+        node = self.ctx.children[self.ctx.iteration - 1]
+
+        if node.is_excepted:
+            return self.exit_codes.ERROR_SUB_PROCESS_EXCEPTED  # pylint: disable=no-member
+
+        if node.is_killed:
+            return self.exit_codes.ERROR_SUB_PROCESS_KILLED  # pylint: disable=no-member
+
+        if node.is_finished_ok:
+
+            # The process was successful so we reset the unhandled failure flag
+            self.ctx.unhandled_failure = False
+
+            exit_code = self._call_sanity_checks(node)
+
+            if isinstance(exit_code, ExitCode):
+                if exit_code.status == 0:
+                    self.report('{}<{}> finished successfully, but sanity check failed, restarting'.format(
+                        self.ctx.process_name, node.pk))
+                else:
+                    self.report('{}<{}> finished successfully, but sanity check detected unrecoverable problem'.format(
+                        self.ctx.process_name, node.pk))
+
+        else:
+
+            try:
+                exit_code = self._call_error_handlers(node)
```

Comment:
Shouldn't this happen _before_ the sanity checks? If I understand this code right, if the error handler succeeds (i.e., it becomes _no longer an error_), the sanity checks will not be called.